### ISSUE TITLE
[bug]: unable to restore backup from FreePBX13 - "Undefined property.... Legacy::$database"

### ISSUE THREAD

**Issue by dairyeng** [FreePBX/issue-tracker#103]
### FreePBX Version

FreePBX 17

### Issue Description

Alright I think I got all the correct info into the freaking ticket this time(pls change the default values to FreePBX17 so you don't keep getting issues for 17 tagged at ver16....)

It really shouldn't be this hard to get a backup restored.............. I follow directions and it doesn't work. The closest thing I've been able to google leads me to believe I have to install php5.6, run the restore, and the restore whatever version of php FreePBX17 is using? That's... BS, guys. Fix this.

In the meantime, please provide workaround steps to get this fixed:

Determining backup file type...type is legacy
Legacy CDR Restore Option: 0
Starting restore job with file: /var/spool/asterisk/backup/uploads/20240420-141451-1713622491-75204414.tgz
Extracting backup...
Loading manifest to memory
Loading astdb to memory
Parsing out SQL tables. This may take a moment depending on backup size.

In Legacy.php line 468:

  Undefined property: FreePBX\modules\Backup\Handlers\Restore\Legacy::$databa
  se

### Operating Environment

Debian 12 fresh install and used the script to install FreePBX17

### Relevant log output

```shell
Determining backup file type...type is legacy
Legacy CDR Restore Option: 0
Starting restore job with file: /var/spool/asterisk/backup/uploads/20240420-141451-1713622491-75204414.tgz
Extracting backup...
Loading manifest to memory
Loading astdb to memory
Parsing out SQL tables. This may take a moment depending on backup size.

In Legacy.php line 468:

  Undefined property: FreePBX\modules\Backup\Handlers\Restore\Legacy::$databa
  se
```

**Comment by kapilgupta01** [comment]
hi @dairyeng  thanks for reporting this issue.
Can you please try with below release ?

Backup v17.0.5.15

_fwconsole ma downloadinstall backup --tag=17.0.5.15_

Regards
Kapil

**Comment by dairyeng** [comment]
Hi there,

Following your instructions got me further, thank you!
Now I get this error about a driver:

File named: /tmp/backup/ace7d1a6-a6a2-4f96-8687-0e31192d3d1b/mysql-3.sql
Detected file /tmp/backup/ace7d1a6-a6a2-4f96-8687-0e31192d3d1b/mysql-3.sql as the PBX (Asterisk) database. Attempting restore
Loading supplied database file mysql-3

In Legacy.php line 313:
                         
  could not find driver  

Googling lead me to try installing php7.0-sqlite3, but doing so and rebooting the server didn't seem to resolve it.

**Comment by dairyeng** [comment]
Alright I got it. I needed to install the current version of php sqlite3 which was 8.2
The commands were:
sudo apt update && sudo apt upgrade
sudo apt -y install php8.2-sqlite3
/etc/init.d/apache2 restart
I then retried a backup from the admin webpage. Looking good until what I /think/ was the end? I got this message:

#13 /var/www/html/admin/libraries/Composer/vendor/symfony/console/Application.php(168): Symfony\Component\Console\Application->doRun()
#14 /var/lib/asterisk/bin/fwconsole(163): Symfony\Component\Console\Application->run()
#15 {main}
Restore processing for modules are finished successfully
npm WARN deprecated is-buffer@1.0.2: This version of 'is-buffer' is out-of-date. You must update to v1.1.6 or newer
npm WARN deprecated nomnom@1.8.1: Package no longer supported. Contact support@npmjs.com for more info.

In Common.php line 309:
                             
  Undefined array key "etc"  

I close the backup message screen and hit "Apply Config" and then check around the config and see if it actually restored stuff... I now have some messages complaining about invalid destinations on my Inbound Routes that weren't there before so it seems like it did /something/ - however the sip extensions don't seem to be there (except for my 2 test extensions I added a few days ago - they're still there)

**Comment by kapilgupta01** [comment]
may be you can try to restore one more time to ensure all the data got restored properly.

**Comment by kapilgupta01** [comment]
also update your backup module by executing  "fwconsole ma downloadinstall backup --tag=17.0.5.16"  for the below error and try to restore again.  

`In Common.php line 309:

Undefined array key "etc"
`

**Comment by dairyeng** [comment]
I ended up just spending a little time and making the bulk ext export csv's data match the FreePBX17 bulk import csv template and pulled it in that way. I was able to verify our recordings and voicemails, tftpboot folder, etc. all transferred successfully. I've been able to successfully connect my home Yealink deskphone to the work PBX server via DDNS and make a test call to my cell phone with audio on both ends, so I think the bulk of things are working now, thank you!!

I guess my final things to do are reset the new server's IP address to the old's and then push out the updated config file to our phones. I've seen some bug reports about the firewall and resetting the PBX's IP address so I'll go give them a look over before pulling the trigger on that one.

I guess my final question that's got me scratching my head is an error on the dashboard saying:

There are 1 bad destinations
DEST STATUS: CUSTOM
Set CallerID: (our company name redacted)

I was able to clear out the other ones by manually recreating the IVR, hangup, time conditions, destinations - but this last one just complains about our CallerID in one of the destinations and I don't see anything wrong with them or which one is the problem.

**Comment by dairyeng** [comment]
Actually, another thing, but maybe I should make a new thread for it:

when clicking on "parking" I get this error message (screenshot attached)
![Screenshot 2024-04-22 235151](https://github.com/FreePBX/issue-tracker/assets/50673605/65993513-4034-46a2-bb69-312afbf39b69)

Whoops \ Exception \ ErrorException (E_WARNING)
Undefined variable $parkedmusicclass

**Comment by kapilgupta01** [comment]
Hi @dairyeng for your parking issue please try with

_fwconsole ma downloadinstall parking --tag=17.0.2.1_

**Comment by dairyeng** [comment]
Awesome that fixed it! Thanks so much!
I've noticed the Phone Restart option is missing from this version and trying to isntall it says it conflicts with the Endpoint Manager - when I try to go to the Endpoint Manager I get a similar error to what I was getting with Parking(screenshot attached), does Endpoint Manager need an update too? Is there an easy way to force update everything?
![Screenshot 2024-04-23 000407](https://github.com/FreePBX/issue-tracker/assets/50673605/d7fc9dd2-1568-4cad-ba1e-8c2617465e3c)

**Comment by kapilgupta01** [comment]
Hi @dairyeng reinstall endpoint manager please.

_fwconsole ma downloadinstall endpoint --edge_

**Comment by dairyeng** [comment]
That worked thanks! Though, now I realize it's only for phones we don't use - all ours are Yealink and we need the Bulk Phone Restart app that says it's conflicting with it. I've removed Endpoint Manager and I'm trying to install "restart" but it's giving me an error about it being depreciated. --force isn't working on it :C 
![Screenshot 2024-04-23 010101](https://github.com/FreePBX/issue-tracker/assets/50673605/d6300c68-b807-4aef-a1e5-ca2194100adf)

**Comment by kapilgupta01** [comment]
Hi @dairyeng Restart module is not supported by us so we are not going to do any fixes there and also thats 16 version which supports only php 7.4
Request you to please use endpoint manager to configure/provisioned your phones.

Thank you for using FreePBX

**Comment by dairyeng** [comment]
They're Yealink Chan_SIP phones - those are manageable through Endpoint Manager?

I've noticed that the dashboard doesn't seem to show my tests phones connected or the trunks connected when I have them enabled - is there a bug or does it just take time for the graphs to actually show any info?

**Comment by dairyeng** [comment]
Another thing - I've noticed that after successfully manually uploading some config files to the tftpboot folder via Filezilla it seems like the folder or files are write protected? I've tried resetting the ftp username and password via the admin page but I'm still only able to connect as the specified user and not upload/delete/overwrite files like I was earlier. It seemed to happen between removing Endpoint Manager, trying to install Reboot, failing to do that, and reinstalling Endpoint Manager and its dependencies.

**Comment by dairyeng** [comment]
Ehhh scratch what I said earlier. I shut off the old phone server and switched the new server's ip address to the old servers IP for the firewall etc and realized that my phones and test calls had connected to the old PBX via the new PBX's DDNS address. Witht he old server off I realize the phones aren't actually connecting to the new PBX.

**Comment by dairyeng** [comment]
okay so I went ahead and started over with a blank Debian 12 and FreePBX17 install with the newest versions of things, did no manual setup and tried running the restore again fresh.

It looks like it got through a lot smoother than last time, but it did notify me of a few errors so I'll go ahead and report them:
I'll check around the config now and see what all it's brought over correctly.

Skipping directory www
Reloading...... DONE
Running Post Restore Hooks.. Please note that hook will restart httpd service so please refresh your page (using new ports) 
Running Post Restore Hooks DONE
Finished
File named: /tmp/backup/9316ec32-0c08-40b8-afb7-88083ce55942/mysql-4.sql
There were errors during the restore process
	Undefined array key "amd" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "builtin" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "bulkhandler" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "calendar" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "callaccounting" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "configedit" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "warmspareenabled" on line 191 of file /var/www/html/admin/modules/core/Restore.php
	Undefined array key "module" on line 373 of file /var/www/html/admin/modules/backup/RestoreBase.php
	Undefined array key "AMPUSERS" on line 15 of file /var/www/html/admin/modules/dictate/Restore.php
	Undefined array key "extensionsettings" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "filestore" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "firewall" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "module" on line 373 of file /var/www/html/admin/modules/backup/RestoreBase.php
	Undefined array key "module" on line 373 of file /var/www/html/admin/modules/backup/RestoreBase.php
	Undefined array key "pbxmfa" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "pm2" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "module" on line 373 of file /var/www/html/admin/modules/backup/RestoreBase.php
	Undefined array key "queuestats" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "module" on line 373 of file /var/www/html/admin/modules/backup/RestoreBase.php
	Undefined array key "sangomaconnect" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "sangomartapi" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "warmspareenabled" on line 177 of file /var/www/html/admin/modules/sipsettings/Restore.php
	Undefined array key "smsplus" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "soundlang" on line 281 of file /var/www/html/admin/modules/backup/Handlers/Restore/Legacy.php
	Undefined array key "warmspareenabled" on line 74 of file /var/www/html/admin/modules/sysadmin/Restore.php
	Undefined array key "module" on line 373 of file /var/www/html/admin/modules/backup/RestoreBase.php
	Undefined array key "module" on line 373 of file /var/www/html/admin/modules/backup/RestoreBase.php
	Running an unsupported version of Asterisk. Supported Asterisk versions: 18, 19, 20, 21. Detected Asterisk version: 13.7.2  on line 30 of file /var/www/html/admin/modules/webrtc/Webrtc.class.php
npm WARN deprecated is-buffer@1.0.2: This version of 'is-buffer' is out-of-date. You must update to v1.1.6 or newer
npm WARN deprecated nomnom@1.8.1: Package no longer supported. Contact support@npmjs.com for more info.

**Comment by dairyeng** [comment]
Hmmm now it can't start Asterisk. Glad I made a backup of the base install before attempting the restore so I can re-try this as updates fixes come available.

I did notice an error in the restore screen about my chan_sip extensions this time - what I'll try tonight is changing to pjsip on the old server, outputting another backup and then switching back so I don't screw up call flows tomorrow, and then try restoring from that to see if it pulls in the extensions.

**Comment by dairyeng** [comment]
I just tried spinning up a FreePBX 13 server and loading my backup into it. 
Something about my backup kills the new servers...... Guess I'm doomed to have to do this all the hard way.
![Screenshot 2024-04-25 192911](https://github.com/FreePBX/issue-tracker/assets/50673605/0bcc89e4-945c-4260-9dd5-b1a6559065b5)

**Comment by kapilgupta01** [comment]
Hi @dairyeng Freepbx 13 is EOL from quite some time now so request you to please either try with 16 or 17.

**Comment by dairyeng** [comment]
This was a test.
I did try with 17(the start of this thread through most of it), and it seemed like every time I restored the backup on 17 it broke the admin site whenever I'd try to browse different things(the trunks page for example). I loaded 13 because it's the same version I'm trying to get off of and where the backup's coming from to test a 13 to 13 restore.
 
My backup breaks the admin portal on version 13 as well.

I think there's something just really wrong with the 13 server the backups are coming from and it's probably just best to rebuild it on a fresh install at this point(this is a server I inherited care of and it /works/ I just can't add/update extensions without introducing the bug I'm trying to upgrade away from to the extensions.). I will ultimately end up doing so on 16 or 17.
